<a href="https://colab.research.google.com/github/Ellizence01/dataviz2025/blob/main/midterm2025_663020587_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

ให้สืบหาฆาตรกร (10 คะแนน) และ ผู้บงการ (5 คะแนน) ในคดี ฆาตรกรรม (murder) ที่เกิดขึ้นในวันที่ 15 มกราคม 2018 ที่เมือง Pandas City โดยเริ่มจากการตรวจสอบ ข้อมูลคดีในไฟล์ crime_scene_report.csv

In [ ]:
import pandas as pd


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
crime = pd.read_csv('/content/drive/MyDrive/midtermdataviz2025/crime_scene_report.csv')
crime.head()

date     type                                        description  \
0  20180115  robbery  A Man Dressed as Spider-Man Is on a Robbery Spree   
1  20180115   murder                  Life? Dont talk to me about life.   
2  20180115   murder  Mama, I killed a man, put a gun against his he...   
3  20180215   murder                         REDACTED REDACTED REDACTED   
4  20180215   murder  Someone killed the guard! He took an arrow to ...   

          city  
0          NYC  
1       Albany  
2         Reno  
3  Pandas City  
4  Pandas City

In [ ]:
crime1 = crime.dropna()
print("DataFrame after removing rows with missing values:")
display(crime1.head())

DataFrame after removing rows with missing values:


date     type                                        description  \
0  20180115  robbery  A Man Dressed as Spider-Man Is on a Robbery Spree   
1  20180115   murder                  Life? Dont talk to me about life.   
2  20180115   murder  Mama, I killed a man, put a gun against his he...   
3  20180215   murder                         REDACTED REDACTED REDACTED   
4  20180215   murder  Someone killed the guard! He took an arrow to ...   

          city  
0          NYC  
1       Albany  
2         Reno  
3  Pandas City  
4  Pandas City

In [ ]:
crime1['date'] = pd.to_datetime(crime1['date'], format='%Y%m%d')
crime1['year'] = crime1['date'].dt.year
crime1['month'] = crime1['date'].dt.month
crime1['day'] = crime1['date'].dt.day


In [ ]:
crime1

date     type                                        description  \
0    2018-01-15  robbery  A Man Dressed as Spider-Man Is on a Robbery Spree   
1    2018-01-15   murder                  Life? Dont talk to me about life.   
2    2018-01-15   murder  Mama, I killed a man, put a gun against his he...   
3    2018-02-15   murder                         REDACTED REDACTED REDACTED   
4    2018-02-15   murder  Someone killed the guard! He took an arrow to ...   
...         ...      ...                                                ...   
1223 2018-04-30  bribery                                                 \n   
1224 2018-04-30    fraud                  ‘Why not?’ said the March Hare.\n   
1225 2018-04-30  assault                                                 \n   
1226 2018-05-01  assault  be NO mistake about it: it was neither more no...   
1227 2018-01-15   murder  Security footage shows that there were 2 witne...   

              city  year  month  day  
0              NYC  2018      1   15  
1           Albany  2018      1   15  
2             Reno  2018      1   15  
3      Pandas City  2018      2   15  
4      Pandas City  2018      2   15  
...            ...   ...    ...  ...  
1223  Garden Grove  2018      4   30  
1224         Houma  2018      4   30  
1225       Fontana  2018      4   30  
1226       Trenton  2018      5    1  
1227   Pandas City  2018      1   15  

[1228 rows x 7 columns]

In [32]:
#ได้ข้อมูลมาว่า ภาพจากกล้องวงจรปิดแสดงให้เห็นว่ามีพยาน 2 คน พยานคนแรกอาศัยอยู่ที่บ้านหลังสุดท้ายบนถนน "Northwestern Dr" ส่วนพยานคนที่สองชื่อแอนนาเบล อาศัยอยู่ที่ไหนสักแห่งบนถนน "Franklin Ave"
crime_filter = crime1[(crime1['date'] == '2018-01-15') & (crime1['city'] == 'Pandas City')&(crime1['type'] == 'murder')]
crime_filter

date    type                                        description  \
1227 2018-01-15  murder  Security footage shows that there were 2 witne...   

             city  year  month  day  
1227  Pandas City  2018      1   15

In [35]:
#หาพยาน
witness = crime = pd.read_csv('/content/drive/MyDrive/midtermdataviz2025/person.csv')


In [37]:
witness_cleaned = witness.dropna()
witness_cleaned

id                 name  license_id  address_number  \
0      10000   Christoper Peteuil      993845             624   
1      10007  Kourtney Calderwood      861794            2791   
2      10010            Muoi Cary      385336             741   
3      10016          Era Moselle      431897            1987   
4      10025         Trena Hornby      550890             276   
...      ...                  ...         ...             ...   
10006  99936          Luba Benser      274427             680   
10007  99941      Roxana Mckimley      975942            1613   
10008  99965      Cherie Zeimantz      287627            3661   
10009  99982          Allen Cruse      251350            3126   
10010  99990         Vance Hunten      830407            3056   

      address_street_name        ssn  
0            Bankhall Ave  747714076  
1           Gustavus Blvd  477972044  
2         Northwestern Dr  828638512  
3           Wood Glade St  614621061  
4           Daws Hill Way  223877684  
...                   ...        ...  
10006        Carnage Blvd  685095054  
10007             Gate St  512136801  
10008       The Water Ave  362877324  
10009           N Jean Dr  348734531  
10010       Lancefield St  896677562  

[10011 rows x 6 columns]

In [52]:
witness_filter = witness_cleaned[witness_cleaned['address_street_name'].isin(['Northwestern Dr', 'Franklin Ave'])]
display(witness_filter)

id            name  license_id  address_number address_street_name  \
2     10010       Muoi Cary      385336             741     Northwestern Dr   
235   12207  Wilmer Wolever      509484             139        Franklin Ave   
282   12711  Norman Apolito      667757             599     Northwestern Dr   
499   14887  Morty Schapiro      118009            4919     Northwestern Dr   
528   15171    Weldon Penso      336999             311     Northwestern Dr   
...     ...             ...         ...             ...                 ...   
9609  96595   Coretta Cubie      303645            3631     Northwestern Dr   
9756  97913  Cameron Dilick      971988            2954        Franklin Ave   
9841  98593     Jonah Toner      578982            2947     Northwestern Dr   
9864  98744   Jordan Myntti      256116            3104        Franklin Ave   
9991  99826    Ivy Kazarian      994470             892     Northwestern Dr   

            ssn  
2     828638512  
235   636825374  
282   778264744  
499   111564949  
528   131379495  
...         ...  
9609  378403829  
9756  665147939  
9841  745853548  
9864  876030104  
9991  208359813  

[95 rows x 6 columns]

In [94]:
northwestern_witness = witness_filter[witness_filter['address_street_name'] == 'Northwestern Dr']
northwestern_witness

id                  name  license_id  address_number  \
2     10010             Muoi Cary      385336             741   
282   12711        Norman Apolito      667757             599   
499   14887        Morty Schapiro      118009            4919   
528   15171          Weldon Penso      336999             311   
811   17729        Lasonya Wildey      439686            3824   
904   18376              Josh Shi      653712            1091   
1021  19420           Cody Schiel      890431            3524   
1338  22239         Dusty Sigafus      710517            1125   
1426  23044          Val Portlock      924989            3143   
1522  23960  Kristopher Lagerberg      658777            1392   
1686  25615         Olevia Morena      388477             812   
1812  26758      Angelena Billman      942074             629   
1981  28360        Rashad Cascone      924517            3212   
2131  29608       Willetta Lurvey      800568            1028   
2535  33415         Homer Simpson      108392             981   
2620  34155      Allyson Lazenson      814209            2840   
2635  34352    Vincenza Burkhardt      436470            2512   
2871  36378        Freddie Ellzey      267882            3449   
3237  39688            Abe Roeker      537445             689   
3316  40336       Fredia Billeter      251082             404   
3368  40760          Renata Faver      976650             745   
3743  44004           Alison Eska      327797            2951   
3795  44542        Ivory Mcgrevey      873842             897   
4595  51114     Christena Saffell      814284            3055   
4643  51552           Bonita Knie      764124            1726   
4806  53076           Boris Bijou      664914            3327   
4871  53699      Christian Pagano      628433            2585   
4886  53890        Sophie Tiberio      957671            3755   
4956  54564         Sammie Sengun      217286            2590   
5024  55130        Shiloh Huntoon      472118             951   
5450  59307          Cyril Yongue      753524            2903   
6476  68690            Yer Modest      852404            3046   
6698  70485     Andera Kostelecky      354822            1252   
6865  71924        Laurie Scronce      774116            1418   
6877  72076     Courtney Bordeaux      324614            1707   
7040  73368        Torie Thalmann      773862            3697   
7296  75484          Del Tacderen      447887             607   
7888  80921       Donovan Yackley      222395            2386   
7902  81052     Calvin Czelusniak      471483            1516   
7993  81885         Kurtis Masuda      934201            2108   
8369  85280             Donn Duin      569122             540   
8608  87456   Leonora Wolfsberger      215868            3483   
8720  88423           Teri Ehrich      510876            1879   
8888  89906       Kinsey Erickson      510019             309   
8939  90397     Dionna Kranwinkle      797260            2678   
9272  93509        Emmitt Aceuedo      916706            3491   
9569  96178           Galina Khuu      450248            2183   
9609  96595         Coretta Cubie      303645            3631   
9841  98593           Jonah Toner      578982            2947   
9991  99826          Ivy Kazarian      994470             892   

     address_street_name        ssn  
2        Northwestern Dr  828638512  
282      Northwestern Dr  778264744  
499      Northwestern Dr  111564949  
528      Northwestern Dr  131379495  
811      Northwestern Dr  917817122  
904      Northwestern Dr  193899001  
1021     Northwestern Dr  947110049  
1338     Northwestern Dr  724386723  
1426     Northwestern Dr  100593316  
1522     Northwestern Dr  492912529  
1686     Northwestern Dr  962831331  
1812     Northwestern Dr  341854756  
1981     Northwestern Dr  838907424  
2131     Northwestern Dr  843751378  
2535     Northwestern Dr  871392019  
2620     Northwestern Dr  991150469  
2635     Northwestern Dr  197913046  
2871     

In [97]:
#พยานคนแรกอาศัยอยู่ที่บ้านหลังสุดท้ายบนถนน "Northwestern Dr"
northwestern_witness_sorted = northwestern_witness.sort_values(by='address_number', ascending=False)
northwestern_witness_sorted

id                  name  license_id  address_number  \
499   14887        Morty Schapiro      118009            4919   
811   17729        Lasonya Wildey      439686            3824   
4886  53890        Sophie Tiberio      957671            3755   
7040  73368        Torie Thalmann      773862            3697   
9609  96595         Coretta Cubie      303645            3631   
1021  19420           Cody Schiel      890431            3524   
9272  93509        Emmitt Aceuedo      916706            3491   
8608  87456   Leonora Wolfsberger      215868            3483   
2871  36378        Freddie Ellzey      267882            3449   
4806  53076           Boris Bijou      664914            3327   
1981  28360        Rashad Cascone      924517            3212   
1426  23044          Val Portlock      924989            3143   
4595  51114     Christena Saffell      814284            3055   
6476  68690            Yer Modest      852404            3046   
3743  44004           Alison Eska      327797            2951   
9841  98593           Jonah Toner      578982            2947   
5450  59307          Cyril Yongue      753524            2903   
2620  34155      Allyson Lazenson      814209            2840   
8939  90397     Dionna Kranwinkle      797260            2678   
4956  54564         Sammie Sengun      217286            2590   
4871  53699      Christian Pagano      628433            2585   
2635  34352    Vincenza Burkhardt      436470            2512   
7888  80921       Donovan Yackley      222395            2386   
9569  96178           Galina Khuu      450248            2183   
7993  81885         Kurtis Masuda      934201            2108   
8720  88423           Teri Ehrich      510876            1879   
4643  51552           Bonita Knie      764124            1726   
6877  72076     Courtney Bordeaux      324614            1707   
7902  81052     Calvin Czelusniak      471483            1516   
6865  71924        Laurie Scronce      774116            1418   
1522  23960  Kristopher Lagerberg      658777            1392   
6698  70485     Andera Kostelecky      354822            1252   
1338  22239         Dusty Sigafus      710517            1125   
904   18376              Josh Shi      653712            1091   
2131  29608       Willetta Lurvey      800568            1028   
2535  33415         Homer Simpson      108392             981   
5024  55130        Shiloh Huntoon      472118             951   
3795  44542        Ivory Mcgrevey      873842             897   
9991  99826          Ivy Kazarian      994470             892   
1686  25615         Olevia Morena      388477             812   
3368  40760          Renata Faver      976650             745   
2     10010             Muoi Cary      385336             741   
3237  39688            Abe Roeker      537445             689   
1812  26758      Angelena Billman      942074             629   
7296  75484          Del Tacderen      447887             607   
282   12711        Norman Apolito      667757             599   
8369  85280             Donn Duin      569122             540   
3316  40336       Fredia Billeter      251082             404   
528   15171          Weldon Penso      336999             311   
8888  89906       Kinsey Erickson      510019             309   

     address_street_name        ssn  
499      Northwestern Dr  111564949  
811      Northwestern Dr  917817122  
4886     Northwestern Dr  442830147  
7040     Northwestern Dr  341559436  
9609     Northwestern Dr  378403829  
1021     Northwestern Dr  947110049  
9272     Northwestern Dr  979073160  
8608     Northwestern Dr  565203106  
2871     Northwestern Dr  474117596  
4806     Northwestern Dr  401191868  
1981     Northwestern Dr  838907424  
1426     Northwestern Dr  100593316  
4595     Northwestern Dr  247657547  
6476     Northwestern Dr  419238039  
3743     Northwestern Dr  306417867  
9841     Northwestern Dr  745853548  
5450     Northwestern Dr  866057839  
2620     

In [85]:
#หาแอนนาเบล
franklinave_witness = witness_filter[witness_filter['address_street_name'] == 'Franklin Ave']
franklinave_witness

id                 name  license_id  address_number  \
235   12207       Wilmer Wolever      509484             139   
665   16371       Annabel Miller      490173             103   
807   17683        Johnnie Schee      968887            1277   
936   18651        Carleen Etoll      356746              22   
1377  22636       Zachary Ybarbo      768359             785   
1608  24737           Gema Nantz      273410            3968   
2236  30654      Clarita Rickels      418084            2254   
2427  32264       Shelby Dezeeuw      735415            1391   
2584  33793         Amado Mattan      161915              99   
2661  34592    Cordell Lindamood      592762            3657   
2864  36328         Domenic Laun      971080            3809   
2877  36400         Noe Pancoast      895755            3443   
3009  37616     Katelyn Schwerin      430143             332   
4086  46827         Bev Billiter      643591            2316   
4708  52059        Jordan Pelton      686854            2786   
4709  52075             Ezra Phy      928109             384   
4714  52121       Yelena Whitler      427830            3185   
4990  54817          Guy Bustard      213929             591   
4998  54896        Patty Wamsley      688488            1097   
5142  56142       William Shiver      741336            1926   
5233  57051        Lavette Koehl      137365            2284   
5497  59746        Ronny Gumbert      469722            3197   
5641  60944        Ricki Bidding      443364            2072   
5653  61001      Laurine Bousman      197150             247   
5709  61437     Colette Hollomon      661377             453   
5810  62292       Isaiah Holsten      913856             747   
5829  62520      Mariah Lepetich      450447            3086   
6326  67292       Renita Roperto      801041             625   
6597  69719        Hope Arzabala      311724             841   
6641  70050   Whitney Herkenratt      138462            3445   
7300  75510  Sebastian Ramnarase      333188            1127   
7357  76128         Ressie Razze      633497            1199   
7574  78170          Graig Summy      192325            3387   
7610  78408         Dante Eschen      218173            3447   
7637  78658        Blake Chrones      310242            2014   
7658  78830        Candice Train      814779            3255   
8120  83003       Wilmer Casella      672050            3564   
8203  83754          Maria Walsh      201334            2716   
8302  84531       Edgar Mendieta      315790            3881   
8624  87521          Moses Ikerd      638991            2283   
8783  88966            Raul Eads      570927            2066   
9322  93947        Omer Andreoni      605718              13   
9462  95119            Hong Lisa      825828             375   
9756  97913       Cameron Dilick      971988            2954   
9864  98744        Jordan Myntti      256116            3104   

     address_street_name        ssn  
235         Franklin Ave  636825374  
665         Franklin Ave  318771143  
807         Franklin Ave  815977821  
936         Franklin Ave  193369255  
1377        Franklin Ave  285346605  
1608        Franklin Ave  180545802  
2236        Franklin Ave  714941023  
2427        Franklin Ave  143197463  
2584        Franklin Ave  125205748  
2661        Franklin Ave  509890333  
2864        Franklin Ave  825329637  
2877        Franklin Ave  732212073  
3009        Franklin Ave  392365033  
4086        Franklin Ave  940192302  
4708        Franklin Ave  676224073  
4709        Franklin Ave  176404498  
4714        Franklin Ave  115106155  
4990        Franklin Ave  820850670  
4998        Franklin Ave  725272037  
5142        Franklin Ave  219913461  
5233        Franklin Ave  776046786  
5497        Franklin Ave  561067477  
5641        Franklin Ave  614976330  
5653        Franklin Ave  431360364  
5709        Franklin Ave  314956038  
5810        Franklin Ave  995368430  
5829        Franklin Ave  797745632  


In [62]:
#พยาน Annabel Miller
annabel_row = franklinave_witness.loc[665]
display(annabel_row)

id                              16371
name                   Annabel Miller
license_id                     490173
address_number                    103
address_street_name      Franklin Ave
ssn                         318771143
Name: 665, dtype: object

In [95]:
#พยาน2 Morty Schapiro
someone_row = northwestern_witness.loc[499]
display(someone_row)

id                               14887
name                    Morty Schapiro
license_id                      118009
address_number                    4919
address_street_name    Northwestern Dr
ssn                          111564949
Name: 499, dtype: object

In [64]:
interview1=pd.read_csv('/content/drive/MyDrive/midtermdataviz2025/interview.csv')
interview1

person_id                                         transcript
0         28508                ‘I deny it!’ said the March Hare.\n
1         63713                                                 \n
2         86208      way, and the whole party swam to the shore.\n
3         35267  lessons in here? Why, there’s hardly room for ...
4         33856                                                 \n
...         ...                                                ...
4986      37357  Alice did not wish to offend the Dormouse agai...
4987      10206  time,’ she said, ‘than waste it in asking ridd...
4988      14887  I heard a gunshot and then saw a man run out. ...
4989      16371  I saw the murder happen, and I recognized the ...
4990      67318  I was hired by a woman with a lot of money. I ...

[4991 rows x 2 columns]

In [66]:
interview_cleaned = interview1.dropna()
interview_cleaned

person_id                                         transcript
0         28508                ‘I deny it!’ said the March Hare.\n
1         63713                                                 \n
2         86208      way, and the whole party swam to the shore.\n
3         35267  lessons in here? Why, there’s hardly room for ...
4         33856                                                 \n
...         ...                                                ...
4986      37357  Alice did not wish to offend the Dormouse agai...
4987      10206  time,’ she said, ‘than waste it in asking ridd...
4988      14887  I heard a gunshot and then saw a man run out. ...
4989      16371  I saw the murder happen, and I recognized the ...
4990      67318  I was hired by a woman with a lot of money. I ...

[4991 rows x 2 columns]

In [68]:
#ปากคำพยาน1 "ฉันเห็นเหตุการณ์ฆาตกรรมเกิดขึ้น และฉันจำฆาตกรได้จากโรงยิมของฉันในขณะที่ฉันกำลังออกกำลังกายเมื่อวันที่ 9 มกราคมที่ผ่านมา"
annabel_interview = interview_cleaned[interview_cleaned['person_id'] == 16371]
display(annabel_interview)

person_id                                         transcript
4989      16371  I saw the murder happen, and I recognized the ...

In [96]:
#ปากคำพยาน2 "ฉันได้ยินเสียงปืน แล้วก็เห็นผู้ชายคนหนึ่งวิ่งออกมา เขาถือกระเป๋า "Get Fit Now Gym" หมายเลขสมาชิกบนกระเป๋าขึ้นต้นด้วย "48Z"
#มีแต่สมาชิกระดับ Gold เท่านั้นที่จะมีกระเป๋าแบบนั้น ชายคนนั้นขึ้นรถที่มีป้ายทะเบียน "H42W"
morty_interview = interview_cleaned[interview_cleaned['person_id'] == 14887]
display(morty_interview)

person_id                                         transcript
4988      14887  I heard a gunshot and then saw a man run out. ...

In [74]:
#ดูว่าใครเข้ายิมวันที่ 9 บ้าง
get_fit = pd.read_csv('/content/drive/MyDrive/midtermdataviz2025/get_fit_now_check_in.csv')

In [75]:
get_fit_cleaned = get_fit.dropna()
get_fit_cleaned

membership_id  check_in_date  check_in_time  check_out_time
0            NL318       20180212            329             365
1            NL318       20170811            469             920
2            NL318       20180429            506             554
3            NL318       20180128            124             759
4            NL318       20171027            418            1019
...            ...            ...            ...             ...
2698         4KB72       20170422           1016            1114
2699         4KB72       20170630            408             885
2700         48Z7A       20180109           1600            1730
2701         48Z55       20180109           1530            1700
2702         90081       20180109           1600            1700

[2703 rows x 4 columns]

In [76]:
get_fit_with_date_parts = get_fit_cleaned.copy()
get_fit_with_date_parts['check_in_date'] = pd.to_datetime(get_fit_with_date_parts['check_in_date'], format='%Y%m%d')
get_fit_with_date_parts['year'] = get_fit_with_date_parts['check_in_date'].dt.year
get_fit_with_date_parts['month'] = get_fit_with_date_parts['check_in_date'].dt.month
get_fit_with_date_parts['day'] = get_fit_with_date_parts['check_in_date'].dt.day
get_fit_with_date_parts

membership_id check_in_date  check_in_time  check_out_time  year  month  \
0            NL318    2018-02-12            329             365  2018      2   
1            NL318    2017-08-11            469             920  2017      8   
2            NL318    2018-04-29            506             554  2018      4   
3            NL318    2018-01-28            124             759  2018      1   
4            NL318    2017-10-27            418            1019  2017     10   
...            ...           ...            ...             ...   ...    ...   
2698         4KB72    2017-04-22           1016            1114  2017      4   
2699         4KB72    2017-06-30            408             885  2017      6   
2700         48Z7A    2018-01-09           1600            1730  2018      1   
2701         48Z55    2018-01-09           1530            1700  2018      1   
2702         90081    2018-01-09           1600            1700  2018      1   

      day  
0      12  
1      11  
2      29  
3      28  
4      27  
...   ...  
2698   22  
2699   30  
2700    9  
2701    9  
2702    9  

[2703 rows x 7 columns]

In [83]:
#คนที่เข้ายิมวันเดียวกัน
get_fit_jan9_2018 = get_fit_with_date_parts[(get_fit_with_date_parts['year'] == 2018) & (get_fit_with_date_parts['month'] == 1) & (get_fit_with_date_parts['day'] == 9)]
get_fit_jan9_2018

membership_id check_in_date  check_in_time  check_out_time  year  month  \
113          X0643    2018-01-09            957            1164  2018      1   
889          UK1F2    2018-01-09            344             518  2018      1   
1329         XTE42    2018-01-09            486            1124  2018      1   
1446         1AE2H    2018-01-09            461             944  2018      1   
1703         6LSTG    2018-01-09            399             515  2018      1   
2137         7MWHJ    2018-01-09            273             885  2018      1   
2435         GE5Q8    2018-01-09            367             959  2018      1   
2700         48Z7A    2018-01-09           1600            1730  2018      1   
2701         48Z55    2018-01-09           1530            1700  2018      1   
2702         90081    2018-01-09           1600            1700  2018      1   

      day  
113     9  
889     9  
1329    9  
1446    9  
1703    9  
2137    9  
2435    9  
2700    9  
2701    9  
2702    9

In [80]:
#ดูว่าแต่ละ member คือใคร
get_member = pd.read_csv('/content/drive/MyDrive/midtermdataviz2025/get_fit_now_member.csv')
get_member

id  person_id                name  membership_start_date  \
0    NL318      65076     Everette Koepke               20170926   
1    AOE21      39426        Noe Locascio               20171005   
2    2PN28      63823  Jeromy Heitschmidt               20180215   
3    0YJ24      80651      Waneta Wellard               20171206   
4    3A08L      32858        Mei Bianchin               20170401   
..     ...        ...                 ...                    ...   
179  2V137      41693      Wendell Dulany               20171219   
180  4KB72      79110          Emile Hege               20170522   
181  48Z7A      28819        Joe Germuska               20160305   
182  48Z55      67318       Jeremy Bowers               20160101   
183  90081      16371      Annabel Miller               20160208   

    membership_status  
0                gold  
1             regular  
2              silver  
3                gold  
4              silver  
..                ...  
179            silver  
180           regular  
181              gold  
182              gold  
183              gold  

[184 rows x 5 columns]

In [98]:
merged_members = pd.merge(get_fit_jan9_2018, get_member[['id', 'person_id', 'name', 'membership_status']], left_on='membership_id', right_on='id', how='inner')
display(merged_members[['person_id', 'name', 'membership_status']])

person_id               name membership_status
0      15247     Shondra Ledlow            silver
1      28073   Zackary Cabotage            silver
2      55662     Sarita Bartosh              gold
3      10815   Adriane Pelligra            silver
4      83186      Burton Grippe              gold
5      31523  Blossom Crescenzo           regular
6      92736      Carmen Dimick              gold
7      28819       Joe Germuska              gold
8      67318      Jeremy Bowers              gold
9      16371     Annabel Miller              gold

In [101]:
#เลือกเฉพาะ membership_status = gold ตามปากคำพยานคนที่ 2
gold_members = merged_members[merged_members['membership_status'] == 'gold']
gold_members

membership_id check_in_date  check_in_time  check_out_time  year  month  \
2         XTE42    2018-01-09            486            1124  2018      1   
4         6LSTG    2018-01-09            399             515  2018      1   
6         GE5Q8    2018-01-09            367             959  2018      1   
7         48Z7A    2018-01-09           1600            1730  2018      1   
8         48Z55    2018-01-09           1530            1700  2018      1   
9         90081    2018-01-09           1600            1700  2018      1   

   day     id  person_id            name membership_status  
2    9  XTE42      55662  Sarita Bartosh              gold  
4    9  6LSTG      83186   Burton Grippe              gold  
6    9  GE5Q8      92736   Carmen Dimick              gold  
7    9  48Z7A      28819    Joe Germuska              gold  
8    9  48Z55      67318   Jeremy Bowers              gold  
9    9  90081      16371  Annabel Miller              gold

In [105]:
#ตามปากคำพยาน"ฉันได้ยินเสียงปืน แล้วก็เห็นผู้ชายคนหนึ่งวิ่งออกมา เขาถือกระเป๋า "Get Fit Now Gym" หมายเลขสมาชิกบนกระเป๋าขึ้นต้นด้วย "48Z"
#มีแต่สมาชิกระดับ Gold เท่านั้นที่จะมีกระเป๋าแบบนั้น ชายคนนั้นขึ้นรถที่มีป้ายทะเบียน "H42W" , มีผู้ต้องสงสัย 2 คน
suspects = gold_members.loc[[7, 8]]
display(suspects)

membership_id check_in_date  check_in_time  check_out_time  year  month  \
7         48Z7A    2018-01-09           1600            1730  2018      1   
8         48Z55    2018-01-09           1530            1700  2018      1   

   day     id  person_id           name membership_status  
7    9  48Z7A      28819   Joe Germuska              gold  
8    9  48Z55      67318  Jeremy Bowers              gold

In [128]:
#ตรวจดูป้ายทะเบียน
plate_num = pd.read_csv('/content/drive/MyDrive/midtermdataviz2025/drivers_license.csv')
plate_num

id  age  height eye_color hair_color  gender plate_number  \
0      100280   72      57     brown        red    male       P24L4U   
1      100460   63      72     brown      brown  female       XF02T6   
2      101029   62      74     green      green  female       VKY5KR   
3      101198   43      54     amber      brown  female       Y5NZ08   
4      101255   18      79      blue       grey  female       5162Z1   
...       ...  ...     ...       ...        ...     ...          ...   
10002  999923   19      77     amber      black  female       5L0ZI4   
10003  999940   71      61     green      green    male       1B8QN8   
10004  999981   67      69     brown       blue  female       1684K3   
10005  999986   49      58     green       grey    male       F8F64H   
10006  999993   18      63     black      black  female       6UZO2O   

         car_make    car_model  
0           Acura          MDX  
1        Cadillac          SRX  
2           Scion           xB  
3          Nissan        Rogue  
4           Lexus           GS  
...           ...          ...  
10002         GMC  Sierra 3500  
10003  Mitsubishi      Eclipse  
10004  Land Rover          LR2  
10005       Lexus           LS  
10006    Cadillac      DeVille  

[10007 rows x 9 columns]

In [129]:
person_28819_info = witness_cleaned[witness_cleaned['id'] == 67318]
display(person_28819_info[['name', 'license_id']])

name  license_id
6327  Jeremy Bowers      423327

In [131]:
#เอาไอดีของ Jeremy Bowers หาป้ายทะเบียนตรงกับปากคำพยาน
plate_num_423327 = plate_num[plate_num['id'] == 423327]
display(plate_num_423327)

id  age  height eye_color hair_color gender plate_number   car_make  \
3529  423327   30      70     brown      brown   male       0H42W2  Chevrolet   

     car_model  
3529  Spark LS

In [132]:
#หาคนบงการ
facebook_event= pd.read_csv('/content/drive/MyDrive/midtermdataviz2025/facebook_event_checkin.csv')

In [133]:
facebook_event_67318 = facebook_event[facebook_event['person_id'] == 67318]
display(facebook_event_67318)

person_id  event_id              event_name      date
20009      67318      4719  The Funky Grooves Tour  20180115
20010      67318      1143    SQL Symphony Concert  20171206

In [135]:
#ดูคำสัมภาษณ์ผู้ต้องหา
#The Funky Grooves Tour ฉันถูกจ้างโดยผู้หญิงคนหนึ่งที่มีเงินมากมาย ฉันไม่รู้จักชื่อเธอ แต่ฉันรู้ว่าเธอสูงประมาณ 5 ฟุต 5 นิ้ว (65 นิ้ว)
#หรือ 5 ฟุต 7 นิ้ว (67 นิ้ว) เธอมีผมสีแดงและขับรถ Tesla Model S ฉันรู้ว่าเธอไปคอนเสิร์ต SQL Symphony 3 ครั้งในเดือนธันวาคม 2017
jeremy_interview = interview_cleaned[interview_cleaned['person_id'] == 67318]
display(jeremy_interview)

person_id                                         transcript
4990      67318  I was hired by a woman with a lot of money. I ...

In [137]:
facebook_event_count = facebook_event[facebook_event['event_name'] == 'SQL Symphony Concert']
display(facebook_event_count)

person_id  event_id            event_name      date
11         97207      1143  SQL Symphony Concert  20170322
12         97207      1143  SQL Symphony Concert  20180226
109        92433      1143  SQL Symphony Concert  20170821
110        92433      1143  SQL Symphony Concert  20180327
296        88952      1143  SQL Symphony Concert  20170918
...          ...       ...                   ...       ...
20003      24556      1143  SQL Symphony Concert  20171224
20006      99716      1143  SQL Symphony Concert  20171206
20007      99716      1143  SQL Symphony Concert  20171212
20008      99716      1143  SQL Symphony Concert  20171229
20010      67318      1143  SQL Symphony Concert  20171206

[212 rows x 4 columns]

In [138]:
facebook_event_count['date'] = pd.to_datetime(facebook_event_count['date'], format='%Y%m%d')
facebook_event_dec_2017 = facebook_event_count[(facebook_event_count['date'].dt.year == 2017) & (facebook_event_count['date'].dt.month == 12)]
display(facebook_event_dec_2017)

/tmp/ipython-input-3195515485.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  facebook_event_count['date'] = pd.to_datetime(facebook_event_count['date'], format='%Y%m%d')


person_id  event_id            event_name       date
4087       62596      1143  SQL Symphony Concert 2017-12-25
4392       19260      1143  SQL Symphony Concert 2017-12-14
6552       58898      1143  SQL Symphony Concert 2017-12-20
6620       69699      1143  SQL Symphony Concert 2017-12-14
6717       19292      1143  SQL Symphony Concert 2017-12-13
7266       43366      1143  SQL Symphony Concert 2017-12-07
8395       92343      1143  SQL Symphony Concert 2017-12-12
10909      28582      1143  SQL Symphony Concert 2017-12-20
10910      28582      1143  SQL Symphony Concert 2017-12-15
13719      81526      1143  SQL Symphony Concert 2017-12-02
13745      24397      1143  SQL Symphony Concert 2017-12-08
14312      11173      1143  SQL Symphony Concert 2017-12-23
15305      79312      1143  SQL Symphony Concert 2017-12-03
15896      69325      1143  SQL Symphony Concert 2017-12-06
20001      24556      1143  SQL Symphony Concert 2017-12-07
20002      24556      1143  SQL Symphony Concert 2017-12-21
20003      24556      1143  SQL Symphony Concert 2017-12-24
20006      99716      1143  SQL Symphony Concert 2017-12-06
20007      99716      1143  SQL Symphony Concert 2017-12-12
20008      99716      1143  SQL Symphony Concert 2017-12-29
20010      67318      1143  SQL Symphony Concert 2017-12-06

In [140]:
event_counts = facebook_event_dec_2017.groupby('person_id').size().reset_index(name='event_count')
attendees_three_times = event_counts[event_counts['event_count'] == 3]
display(attendees_three_times)

person_id  event_count
4       24556            3
15      99716            3

In [146]:
#คนบงการ
persons_info = witness_cleaned[witness_cleaned['id'].isin([24556, 99716])]
display(persons_info[['id', 'name', 'license_id']])

id              name  license_id
1587  24556       Bryan Pardo      101191
9985  99716  Miranda Priestly      202298

In [149]:
#Miranda Priestly
plate_num_202298 = plate_num[plate_num['id'] == 202298]
display(plate_num_202298)

id  age  height eye_color hair_color  gender plate_number car_make  \
1105  202298   68      66     green        red  female       500123    Tesla   

     car_model  
1105   Model S

ที่สถานการณ์สามารถสรุปได้ว่า:

* **ฆาตกร (Murderer):** Jeremy Bowers (person_id 67318)
    * ข้อมูลจากรายงานสถานที่เกิดเหตุและการสัมภาษณ์พยานระบุว่าฆาตกรเป็นผู้ชายที่วิ่งออกมาจากที่เกิดเหตุในวันที่ 15 มกราคม 2018 ที่ Pandas City และมีกระเป๋า "Get Fit Now Gym" ที่มีหมายเลขสมาชิกขึ้นต้นด้วย "48Z"
    * จากการตรวจสอบข้อมูลสมาชิก Get Fit Now พบผู้ต้องสงสัย 2 คนที่มีหมายเลขสมาชิกขึ้นต้นด้วย "48Z" และมีสถานะเป็น Gold member (ตามคำให้การของพยาน) คือ Joe Germuska (person_id 28819) และ Jeremy Bowers (person_id 67318)
    * การตรวจสอบป้ายทะเบียนรถ "H42W" (ตามคำให้การของพยาน) พบว่าเชื่อมโยงกับ license_id 423327 ซึ่งเป็นของ Jeremy Bowers (person_id 67318)

* **ผู้บงการ (Mastermind):** Miranda Priestly (person_id 99716)
    * จากการสัมภาษณ์ Jeremy Bowers (ผู้ต้องสงสัยว่าเป็นฆาตกร) เขาให้การว่าถูกจ้างโดยผู้หญิงที่มีเงินมากมาย โดยพบที่ งาน The Funky Grooves Tour
    * ข้อมูลการเช็คอินกิจกรรม Facebook แสดงให้เห็นว่ามีบุคคลสองคนคือ Bryan Pardo (person_id 24556) และ Miranda Priestly (person_id 99716) ที่เข้าร่วมคอนเสิร์ต SQL Symphony 3 ครั้งในเดือนธันวาคม 2017 (ตามคำให้การของ Jeremy Bowers)
    * จากคำให้การของ Jeremy Bowers ที่ระบุลักษณะของผู้บงการ (ผู้หญิง ผมสีแดง ขับรถ Tesla Model S) เมื่อตรวจสอบข้อมูลป้ายทะเบียนของ Miranda Priestly (license_id 202298) พบว่าตรงกับลักษณะดังกล่าว (ผู้หญิง ผมสีแดง และขับรถ Tesla Model S)

ดังนั้น ฆาตกรคือ Jeremy Bowers และผู้บงการคือ Miranda Priestly ซึ่งเป็นการจ้างวานฆ่า